app/src/main/java/de/stephanlindauer/criticalmaps/adapter/ChatMessageAdapter.java

```diff
@@ -12,65 +12,77 @@
 
-        return null;
-    }
+        private final DateFormat dateFormatter = DateFormat.getDateTimeInstance(DateFormat.DEFAULT, DateFormat.SHORT, Locale.getDefault());
+        private final Context context;
 
-    private View buildOutgoingMessageView(OutgoingChatMessage currentMessage, LayoutInflater inflater, ViewGroup parent) {
+        public OutgoingChatMessageViewHolder(View itemView) {
+            super(itemView);
+            context = itemView.getContext();
+            ButterKnife.bind(this, itemView);
+        }
 
-        View rowView = inflater.inflate(R.layout.view_outgoing_chatmessage, parent, false);
-
-        TextView labelView = (TextView) rowView.findViewById(R.id.firstLine);
-        TextView valueView = (TextView) rowView.findViewById(R.id.secondLine);
-
-        labelView.setText(R.string.chat_sending);
-        valueView.setText(currentMessage.getMessage());
-
-        return rowView;
+        public void bind(IChatMessage message) {
+            valueView.setText(message.getMessage());
+            if (message instanceof ReceivedChatMessage) {
+                dateFormatter.setTimeZone(TimeZone.getDefault());
+                labelView.setText(TimeToWordStringConverter.getTimeAgo(((ReceivedChatMessage) message).getTimestamp(), context));
+            } else {
+                labelView.setText(R.string.chat_sending);
+            }
+        }
     }
 
-    private View buildReceivedMessageView(ReceivedChatMessage currentMessage, LayoutInflater inflater, ViewGroup parent) {
-
-        View rowView = inflater.inflate(R.layout.view_chatmessage, parent, false);
-
-        TextView labelView = (TextView) rowView.findViewById(R.id.firstLine);
-        TextView valueView = (TextView) rowView.findViewById(R.id.secondLine);
+    @Override
+    public OutgoingChatMessageViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
+        final LayoutInflater inflater = LayoutInflater.from(parent.getContext());
+        final View res;
+        if (viewType == 0) {
+            res = inflater.inflate(R.layout.view_chatmessage, parent, false);
+        } else {
+            res = inflater.inflate(R.layout.view_outgoing_chatmessage, parent, false);
+        }
+        return new OutgoingChatMessageViewHolder(res);
+    }
 
-        DateFormat dateFormatter = DateFormat.getDateTimeInstance(DateFormat.DEFAULT, DateFormat.SHORT, Locale.getDefault());
-        dateFormatter.setTimeZone(TimeZone.getDefault());
+    @Override
+    public void onBindViewHolder(OutgoingChatMessageViewHolder holder, int position) {
+        holder.bind(chatMessages.get(position));
+    }
 
-        labelView.setText(TimeToWordStringConverter.getTimeAgo(currentMessage.getTimestamp(), context));
-        valueView.setText(currentMessage.getMessage());
+    @Override
+    public int getItemViewType(int position) {
+        if (chatMessages.get(position) instanceof ReceivedChatMessage) {
+            return 0;
+        } else {
```

Comment:
@ligi  Are you assuming that the message will be `OutgoingChatMessage` in the `else` part? You would omit a check here for the future?